modules/sharedaddy/sharing.js

```diff
@@ -1,364 +1,474 @@
-			if ( 'https:' === window.location.protocol ) {
-				url = url.replace( /^http:\/\//i, 'https://' );
-			} else {
-				url = url.replace( /^https:\/\//i, 'http://' );
+		do {
+			var matches = matcher.bind( current );
+			if ( matches( sel ) ) {
+				return current;
 			}
+			current = current.parentElement || current.parentNode;
+		} while ( current !== null && current.nodeType === 1 );
 
-			return url;
-		},
-		update_facebook_count: function ( data ) {
-			var url, permalink;
+		return null;
+	}
 
-			if ( ! data ) {
-				return;
-			}
+	// Helper function to iterate over a NodeList
+	// (since IE 11 doesn't have NodeList.prototype.forEach)
+	function forEachNode( list, fn ) {
+		for ( var i = 0; i < list.length; i++ ) {
+			var node = list[ i ];
+			fn( node, i, list );
+		}
+	}
 
-			for ( url in data ) {
-				if (
-					! data.hasOwnProperty( url ) ||
-					! data[ url ].share ||
-					! data[ url ].share.share_count
-				) {
-					continue;
-				}
+	// Helper function to remove a node from the DOM.
+	function removeNode( node ) {
+		if ( node && node.parentNode ) {
+			node.parentNode.removeChild( node );
+		}
+	}
 
-				permalink = WPCOMSharing.get_permalink( url );
+	// Helper functions to show/hide a node, and check its status.
+	function hideNode( node ) {
+		if ( node ) {
+			node.style.display = 'none';
+		}
+	}
 
-				if ( ! ( permalink in WPCOM_sharing_counts ) ) {
-					continue;
-				}
+	function showNode( node ) {
+		if ( node ) {
+			node.style.removeProperty( 'display' );
+		}
+	}
 
-				WPCOMSharing.inject_share_count(
-					'sharing-facebook-' + WPCOM_sharing_counts[ permalink ],
-					data[ url ].share.share_count
-				);
-			}
-		},
-		update_pinterest_count: function ( data ) {
-			if ( 'undefined' !== typeof data.count && data.count * 1 > 0 ) {
-				WPCOMSharing.inject_share_count(
-					'sharing-pinterest-' + WPCOM_sharing_counts[ data.url ],
-					data.count
-				);
-			}
-		},
-		inject_share_count: function ( id, count ) {
-			var $share = jQuery( 'a[data-shared=' + id + '] > span' );
-			$share.find( '.share-count' ).remove();
-			$share.append(
-				'<span class="share-count">' + WPCOMSharing.format_count( count ) + '</span>'
-			);
-		},
-		format_count: function ( count ) {
-			if ( count < 1000 ) {
-				return count;
-			}
-			if ( count >= 1000 && count < 10000 ) {
-				return String( count ).substring( 0, 1 ) + 'K+';
-			}
-			return '10K+';
-		},
-		bump_sharing_count_stat: function ( service ) {
-			new Image().src =
-				document.location.protocol +
-				'//pixel.wp.com/g.gif?v=wpcom-no-pv&x_sharing-count-request=' +
-				service +
-				'&r=' +
-				Math.random();
-		},
+	function isNodeHidden( node ) {
+		return ! node || node.style.display === 'none';
+	}
+
+	// ------------------------------- CLASSES ------------------------------- //
+
+	// Implements a MoreButton class, which controls the lifecycle and behavior
+	// of a "more" button and its dialog.
+	function MoreButton( buttonEl ) {
```

Comment:
If we are going to introduce a prototype, we might as well introduce a class and things will get much simpler.

The hard thing here is IE11. Should we transpile _just_ for that browser?

One thing I'm thinking is... perhaps we do two builds, one which transpiles for evergreen browsers, and one for IE11. The evergreen browsers can use es6 classes natively, whereas IE11 gets the polyfill stuff. 

This will allow us to write much cleaner and more modern javascript, and the holdout can run the polyfills without everyone else getting them.

Might make a nice blog post at some point so that other plugins can emulate this approach.

Reply:
(This discussion is continued on the first comment)